File: 2022/03/03.rb
# frozen_string_literal: true

path = File.expand_path('../inputs/3.txt', __dir__)

input = File.readlines(path, chomp: true)

LOWER_CASE_SCORES = ('a'..'z').each_with_object({}).with_index { |(char, result), index| result[char] = index + 1 }
SCORES = ('A'..'Z').each_with_object(LOWER_CASE_SCORES).with_index { |(char, result), index| result[char] = index + 27 }

part1 = input.map { |items| items.chars.each_slice(items.length / 2).to_a }.sum do |compartment1, compartment2|
  SCORES[(compartment1 & compartment2)[0]]
end

part2 = input.each_slice(3).map { _1.map(&:chars) }.sum do |first, second, third|
  SCORES[(first & second & third)[0]]
end

puts "part 1:  #{part1} "
puts "part 2:  #{part2} "

File: 2022/03/03.py
with open("inputs/3.txt") as file:
    rucksack_items = [list(rucksack.strip()) for rucksack in file]

def part_1():
    priority_sum = 0
    for items in rucksack_items:
        item_number = len(items)
        compartment1 = set(items[:(item_number//2)])
        compartment2 = set(items[(item_number//2):])
        in_common = compartment1.intersection(compartment2).pop()
        if in_common.lower() == in_common:
            priority_sum += ord(in_common) - 96
        else:
            priority_sum += ord(in_common)-38
    return priority_sum

def part_2():
    priority_sum = 0
    for group_start_index in range(0, len(rucksack_items)-2, 3):
        group = [set(items) for items in rucksack_items[group_start_index:group_start_index + 3]]
        in_common = group[0].intersection(group[1]).intersection(group[2]).pop()
        if in_common.lower() == in_common:
            priority_sum += ord(in_common) - 96
        else:
            priority_sum += ord(in_common)-38
    return priority_sum

print(f"Part 1: {part_1()}")
print(f"Part 2: {part_2()}")
